### PR TITLE
Yet another batch of Tril improvements

### LINKED ISSUE #1508: Proposal to shorten Tril commoned node identifiers
In an attempt to make Tril slightly less verbose when dealing with commoned nodes, instead of declaring `@common id="foo"` can we simply shorten it to `@id="foo"` ?

To my eye, it is still easy to distinguish commoned and uncommoned node identifiers.  It also looks more readable when there are a lot of commoned nodes in play (though, arguably, Tril isn't really supposed to favour human readability).

Any thoughts?

@Leonardo2718

### PR DESCRIPTION
**Pull request by Leonardo2718**
This batch of changes makes several improvements to Tril, including:

* A better API for the AST
* Support for hex literals
* Deprecate `@common` syntax in favor of `@id` (issue #1508)
* More documentation 
* More tests

Fixes #1508

### CODE REVIEW

File: fvtest/tril/test/ASTTest.cpp [hunk 1]
```diff
@@ -18,91 +18,185 @@
...
-    ASSERT_EQ(String, v->type);
-    ASSERT_STREQ(stringBaseValue, v->value.str);
+    ASSERT_EQ(ASTValue::String, v->getType());
+    ASSERT_STREQ(stringBaseValue, v->getString());
     ASSERT_EQ(NULL, v->next);
 }
 
+/*
+ * Since the comparison tests verify the behavior of overloaded comparison
+ * operators, they explicitely call the operators inside a ASSERT_{TRUE|FALSE}
```
Comment (0xdaryl) [inline fvtest/tril/test/ASTTest.cpp]: * sp:  explicitly

* "...inside an ASSERT..."

File: fvtest/tril/tril/ast.hpp [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+ * FloatingPoint | floating point values | any type satisfying `std::is_floating_point`
+ * String        | character strings     | `const char*`
+ *
+ * Names for these types are defined in the ASTType enum.
+ *
+ * To aid usability, this struct provides the `get<T>()` template member function
+ * that will return the stored value as type `T` if `T` is compatible with the
+ * type of the stored value.
+ *
+ * For conevenience Integer_t, FloatingPoint_t, and String_t are provided as
```
Comment (0xdaryl) [inline fvtest/tril/tril/ast.hpp]: sp: convenience

File: fvtest/tril/tril/ast.hpp [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+    // aliases for the underlying C++ types used to store the different AST types
+    using Integer_t = uint64_t;
+    using FloatingPoint_t = double;
+    using String_t = const char *;
+
+    /**
+     * STL compatible type traits for working with AST data types
+     *
+     * The `is_*_Compatible` type traits have their `::value` field set to true
+     * if the specified type is compatibile with the named type, false otherwise.
```
Comment (0xdaryl) [inline fvtest/tril/tril/ast.hpp]: sp: compatible

File: fvtest/tril/tril/ast.hpp [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+#include <type_traits>
+#include <assert.h>
+#include <cstring>
+
+/**
+ * @brief The ASTValue struct represents a "value" in the Tril AST
+ *
+ * The struct acts as a variant capable of holding values of multiple types. It
+ * abstracts away the details of how a value is stored and provides an interface
+ * for retrieving a value using an approriate type.
```
Comment (Leonardo2718) [inline fvtest/tril/tril/ast.hpp]: Fixed!

File: fvtest/tril/tril/ast.hpp (hunk 2 above)
Comment (Leonardo2718) [inline fvtest/tril/tril/ast.hpp]: Fixed!

File: fvtest/tril/tril/ast.hpp (hunk 3 above)
Comment (Leonardo2718) [inline fvtest/tril/tril/ast.hpp]: Fixed!

File: fvtest/tril/test/ASTTest.cpp (hunk 1 above)
Comment (Leonardo2718) [inline fvtest/tril/test/ASTTest.cpp]: Fixed!

**Pull request merged**